beacon-chain/operations/voluntaryexits/pool.go

```diff
@@ -0,0 +1,167 @@
+// This pool is used by proposers to insert voluntary exits into new blocks.
+type PoolManager interface {
+	PendingExits() ([]*ethpb.SignedVoluntaryExit, error)
+	ExitsForInclusion(state state.ReadOnlyBeaconState, slot types.Slot) ([]*ethpb.SignedVoluntaryExit, error)
+	InsertVoluntaryExit(exit *ethpb.SignedVoluntaryExit)
+	MarkIncluded(exit *ethpb.SignedVoluntaryExit)
+}
+
+// Pool is a concrete implementation of PoolManager.
+type Pool struct {
+	lock    sync.RWMutex
+	pending doublylinkedlist.List[*ethpb.SignedVoluntaryExit]
+	m       map[types.ValidatorIndex]*doublylinkedlist.Node[*ethpb.SignedVoluntaryExit]
+}
+
+// NewPool returns an initialized pool.
+func NewPool() *Pool {
+	return &Pool{
+		pending: doublylinkedlist.List[*ethpb.SignedVoluntaryExit]{},
+		m:       make(map[types.ValidatorIndex]*doublylinkedlist.Node[*ethpb.SignedVoluntaryExit]),
+	}
+}
+
+// PendingExits returns all objects from the pool.
+func (p *Pool) PendingExits() ([]*ethpb.SignedVoluntaryExit, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	result := make([]*ethpb.SignedVoluntaryExit, p.pending.Len())
+	node := p.pending.First()
+	var err error
+	for i := 0; node != nil; i++ {
+		result[i], err = node.Value()
+		if err != nil {
+			return nil, err
+		}
+		node, err = node.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+// ExitsForInclusion returns objects that are ready for inclusion at the given slot. This method will not
+// return more than the block enforced MaxVoluntaryExits.
+func (p *Pool) ExitsForInclusion(state state.ReadOnlyBeaconState, slot types.Slot) ([]*ethpb.SignedVoluntaryExit, error) {
+	p.lock.RLock()
+	length := int(math.Min(float64(params.BeaconConfig().MaxVoluntaryExits), float64(p.pending.Len())))
+	result := make([]*ethpb.SignedVoluntaryExit, 0, length)
+	node := p.pending.Last()
+	for node != nil && len(result) < length {
+		exit, err := node.Value()
+		if err != nil {
+			p.lock.RUnlock()
+			return nil, err
+		}
+		if exit.Exit.Epoch > slots.ToEpoch(slot) {
+			node, err = node.Prev()
+			if err != nil {
+				p.lock.RUnlock()
+				return nil, err
+			}
+			continue
+		}
+		validator, err := state.ValidatorAtIndexReadOnly(exit.Exit.ValidatorIndex)
+		if err != nil {
+			logrus.WithError(err).Warning("removing invalid exit from pool")
+			// MarkIncluded removes the invalid exit from the pool
+			p.lock.RUnlock()
+			p.MarkIncluded(exit)
+			p.lock.RLock()
+		}
+		if err = blocks.VerifyExitConditions(validator, state.Slot(), exit.Exit); err != nil {
+			logrus.WithError(err).Warning("removing invalid exit from pool")
+			// MarkIncluded removes the invalid exit from the pool
+			p.lock.RUnlock()
+			p.MarkIncluded(exit)
+			p.lock.RLock()
+		}
+		result = append(result, exit)
+		node, err = node.Prev()
+		if err != nil {
+			p.lock.RUnlock()
+			return nil, err
+		}
+	}
+	p.lock.RUnlock()
+	if len(result) == 0 {
+		return result, nil
+	}
+	// We now verify the signatures in batches
```

Comment:
Why do we need to verify the signatures again? we verified them when we included them. I think this is fine because it's safer, but also I believe we need to flush our operations pools at forks, so then the ones that become invalid when forks changes don't really need to be verified again. 

This comment comes from the fact that this function represents a departure from our current model in which `PendingExits` does not check signatures.

Reply:
If we are going to implement a batch exits verification function , it should be in another PR. This PR should be scoped to just changing the design of the pool. It looks like everything follows  how things are done in BlsToExecutionChanges exactly, which is the wrong way to think about it. Exits and BLSChanges have different characteristics and the pool was designed for the latter to handle certain race conditions in mind. This doesn't exist for voluntary exits

Reply:
But we do verify exit signatures currently:
```
if err := blocks.VerifyExitAndSignature(val, head.Slot(), head.Fork(), exit, head.GenesisValidatorsRoot()); err != nil {
			log.WithError(err).Warn("Could not verify exit for block inclusion")
			continue
		}
```
Batch verification simply speeds up the process